src/nupic/engine/RegionImplFactory.cpp

```diff
@@ -48,210 +47,61 @@
 #define expand_and_stringify(x) stringify(x)
 
 namespace nupic {
-// Keys are Python modules and the values are sets of class names for the
-// regions that have been registered in the corresponding module. E.g.
-// pyRegions["nupic.regions.sample_region"] = "SampleRegion"
-static std::map<const std::string, std::set<std::string>> pyRegions;
 
-// Mappings for C++ regions
-static std::map<const std::string, GenericRegisteredRegionImpl *> cppRegions;
+// Mappings for region nodeTypes that map to Class and module
+static std::map<const std::string, std::shared_ptr<RegisteredRegionImpl> > regionTypeMap;
 
 bool initializedRegions = false;
 
-// Allows the user to add custom regions
-void RegionImplFactory::registerPyRegion(const std::string module,
-                                         const std::string className) {
-  // Verify that no regions exist with the same className in any module
-  for (auto pyRegion : pyRegions) {
-    if (pyRegion.second.find(className) != pyRegion.second.end()) {
-      if (pyRegion.first != module) {
-        // New region class name conflicts with previously registered region
-        NTA_THROW << "A pyRegion with name '" << className
-                  << "' already exists. "
-                  << "Unregister the existing region or register the new "
-                     "region using a "
-                  << "different name.";
-      } else {
-        // Same region registered again, ignore
-        return;
-      }
-    }
-  }
 
-  // Module hasn't been added yet
-  if (pyRegions.find(module) == pyRegions.end()) {
-    pyRegions[module] = std::set<std::string>();
-  }
 
-  pyRegions[module].insert(className);
-}
+void RegionImplFactory::registerRegion(const std::string& nodeType, RegisteredRegionImpl *wrapper) {
+  if (regionTypeMap.find(nodeType) != regionTypeMap.end()) {
+	std::shared_ptr<RegisteredRegionImpl>& reg = regionTypeMap[nodeType];
+	if (reg->className() == wrapper->className() && reg->moduleName() == wrapper->moduleName()) {
+		NTA_WARN << "A Region Type already exists with the name '" << nodeType
+				 << "'. Overwriting it...";
+		reg.reset(wrapper);  // replace this impl
+	} else {
+        NTA_THROW << "A region Type with name '" << nodeType
+                  << "' already exists. Class name='" << reg->className()
+                  << "'  Module='" << reg->moduleName() << "'. "
+                  << "Unregister the existing region Type or register the new "
+                     "region Type using a "
+                  << "different name.";
+	}
+  } else {
+    std::shared_ptr<RegisteredRegionImpl> reg(wrapper);
 
-void RegionImplFactory::registerCPPRegion(
```

Comment:
are these changes to Regions part of the PR, or some old (un)merged changes?

Reply:
>` void RegionImplFactory::registerCPPRegion(`
are these changes to Regions part of the PR,

Yes, a very important part of this PR.  Let me explain.
The RegionImplFactory is the only module that contained references from the Network Framework back into the Python code. Without removing these references we would always require linking with the Python library.  I think they were put there by someone that did not understand how the Plug-in thing works.  

Whoever originally designed this code architectured it correctly so that Region implementations could be added but were entirely independent.  The RegionImplFactory is a singleton that holds a list of region implementation registrations that it can use to instantiate a Region impl when a region is correated. Each registration is held in the RegisteredRegionImpl class, indexed by Region type.

But then someone added code to register Python region implementations with a separate list and that broke the Independence.  To fix this, I added a new subclass of the RegisteredRegionImpl class that handles Python (there is already one for C++ implementations).  Now there is only one list of registrations.  This subclass for Python knows how to instantiate a PyBindRegion object which is the base class for all Python region implications and knows how to call back into Python.  As long as this new RegisteredRegionImplPy class is not linked into the nupic_core library then other (C++ only implementions) can link to the nupic_core  without needing to also link in Python.  So the Python related classes are linked into the Python bindings not the nupic_core. See them in bindings/py/src/plugin.  There is more discussion about this in nupic/engine/RegisteredRegionImpl.hpp and the two subclasses.

Are you sorry you asked?....

Reply:
> Are you sorry you asked?....

Not at all! :wave:  I'm happy I learned something new about the inner working of the codebase. 

> Without removing these references we would always require linking with the Python library. I think they were put there by someone that did not understand how the Plug-in thing works. 

Great you fixed that! So now each binding (py,...) is independent and fully optional. 

> Whoever originally designed this code architectured it correctly so that Region implementations could be added but were entirely independent. The RegionImplFactory is a singleton that holds a list of region implementation registrations that it can use to instantiate a Region impl when a region is correated. Each registration is held in the RegisteredRegionImpl class, indexed by Region type. 

> Now there is only one list of registrations. This subclass for Python knows how to instantiate a PyBindRegion object which is the base class for all Python region implications and knows how to call back into Python. As long as this new RegisteredRegionImplPy class is not linked into the nupic_core library then other (C++ only implementions) can link to the nupic_core without needing to also link in Python. So the Python related classes are linked into the Python bindings not the nupic_core. See them in bindings/py/src/plugin.

do you want to add this as a memo to the docs as well?